### ISSUE TITLE
out= resizing (and restriding) behavior is confusing

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#41027]
Today in PyTorch we may resize and restride tensors provided to as arguments to `out`. For example:

```
a = torch.tensor([1, 2, 3])
b = torch.tensor([4, 5, 6])
out = torch.empty(2)
torch.add(a, b, out=out)
: tensor([5., 7., 9.])
```

Here a tensor of two elements is provided as the out argument, but the tensor returned has three elements. 

This behavior is silent and (anecdotally) a frequent cause of user confusion. 

I would like to propose we deprecate the current behavior and implement the following:

- Tensors with no elements passed to the out argument are treated as references, and PyTorch may allocate storage to hold the results of the operation and assign it to the given tensor
- Tensors with more than zero elements cannot be resized or restrided, and we throw an error if there is a size or striding mismatch

The first behavior is a common pattern internally in PyTorch, and is useful when dealing with operators like masked_select where the size of the output is difficult to determine without running the operation. The second behavior is inconsistent with NumPy, which supports broadcasting to out tensors:

```
a = np.array([1])
b = np.array([2])
out = np.empty(3)
np.add(a, b, out=out)
: array([3., 3., 3.])
```

The NumPy behavior may be confusing, however. Also, throwing a runtime error instead of supporting NumPy behavior is an acceptable inconsistency, and it gives PyTorch the option to implement NumPy-like behavior in the future without it being a BC-breaking change. 

@ngimel @VitalyFedyunin @ezyfang @albanD

**Comment by vadimkantorov** [comment]
Maybe sth to consider. There are some cases where it is not non-trivial to calculate the output size exactly: various kinds of convolutions and similar ops where padding is important; torch.cat. One middle ground could be allowing resizing if already allocated flat storage size is greater or equal of what is required.

**Comment by albanD** [comment]
I agree with vadim that the resizing is important in many cases. And in general in the "buffer" use case.
@vadimkantorov one problem that arise, even without allocation is when the given Tensor is a partial view of a storage (due to previous call with out= or because of slicing) and we call resize_(), then the new Tensor will actually overlap with the other slices of the Tensor. Which can lead to very surprising behavior:
```
import torch

a = torch.rand(4)
b = torch.ones(4)

c = a[:2]

print(a)

# No use of a here
torch.add(b, b, out=c)

print(a)
```


That being said, the "buffer" use case is very niche these days, especially with the high quality allocators that we have.
So maybe, forbidding resize for non-zero size is the way to go (with a global flag to still have the old behavior for people that really want it?)

**Comment by mruberry** [comment]
@zdevito Has previously suggested there's a distinction to be made between Tensors and Storages, too, and maybe for some of these scenarios we should support passing Storages to out with different semantics. For example, maybe to support the case of "I know this storage is big enough, please reuse it" we require the user pass an actual storage.

**Comment by vadimkantorov** [comment]
@albanD I experienced bad allocator behavior and fragmentation with much varying time-dimension batch sizes in time convolutional models. these problems go away if you pad e.g. by multiple of 128. Another way to handle this is to have arena allocators that effectively reset all allocations / tensor size bucketing in allocator once every training iteration (like they do in inference engines like onnxruntime). @ngimel at one point suggested it was valid, but maybe too complex / niche for now.

@mruberry That's a nice idea! A bit more towards arena actually!

**Comment by ngimel** [comment]
>(with a global flag to still have the old behavior for people that really want it?)
No global flags please?

**Comment by albanD** [comment]
Yes the fragmentation is an issue. But engineering the `out=` API to avoid fragmentation sounds like doing it backward. We should solve the fragmentation issues in the allocator itself no?
I like the idea of having different behavior for Storage and Tensor as well, but might be tricky to explain to users.

**Comment by vadimkantorov** [comment]
Yeah, I just wanted to point out that allocators are still not "solved" + buffers can be useful + some ops have non-trivial output size computation (so that preparing perfectly matching-sized buffer can be nasty). As we discussed, multiple solutions are possible...

About allocator, I think that first some easy dumping of allocator state: https://github.com/pytorch/pytorch/issues/1529 https://github.com/pytorch/pytorch/issues/29554 (ideally maybe have some way to dump some visualization of memory allocation, maybe some svg? or some example for doing this on pytorch.org?). Then users can easily save and upload these snapshots at OOM. And then we can understand what's the problem. Maybe arenas can be a way forward (especially for inference-like behavior), or maybe some other sorts of allocator hints, or maybe sth else.

**Comment by mruberry** [comment]
Note from offline conversation with @gchanan: let's be v. careful about what the "zero-element tensor" can actually be, and which of its properties are/aren't respected. For example, we probably want to respect its dtype, but do we want to respect the number of dimensions it has?

**Comment by ngimel** [comment]
My take: TensorImpl (i.e. sparse/dense), device, dtype should be respected, similar to what resize_ does, number of dimensions should not be. That prevents us from providing better debugging info to users who actually wanted to compute the correct output size but made a mistake resulting in a tensor with 0 elements, but otherwise that provides a more streamlined interface, zero elements is pretty easy to understand.

**Comment by mruberry** [comment]
> My take: TensorImpl (i.e. sparse/dense), device, dtype should be respected, similar to what resize_ does, number of dimensions should not be. That prevents us from providing better debugging info to users who actually wanted to compute the correct output size but made a mistake resulting in a tensor with 0 elements, but otherwise that provides a more streamlined interface, zero elements is pretty easy to understand.

SGTM. What would you think about requiring zero elements and 1 dim? (Like the tensor returned from torch.empty(0)?)

**Comment by ngimel** [comment]
I thought about it, and this is the way the tensors to be resized are created in ATen (at::empty({0})), but to me requiring 1d seems too strict and does not serve any good purpose. I don't have a strong opinion on it though.

**Comment by xwang233** [comment]
Cross post from #43541 , what should we expect when user passed in a non-contiguous tensor
```python
import torch
print(torch.__version__)

a = torch.arange(5)
b = torch.ones_like(a)

c = torch.add(a, b)
print(c)

d = torch.empty_like(a).unsqueeze(0).transpose(0, 1)
print(d.size(), d.stride())

torch.add(a, b, out=d)
print(d)
print(d.size(), d.stride())
```

output
```
tensor([1, 2, 3, 4, 5])
torch.Size([5, 1]) (1, 5)
out_tensor.py:13: UserWarning: An output with one or more elements was resized since it had shape [5, 1],
which does not match the required output shape [5].This behavior is deprecated, and in a future PyTorch release
outputs will not be resized unless they have zero elements. You can explicitly reuse an out tensor t by resizing it,
inplace, to zero elements with t.resize_(0). (Triggered internally at  ../aten/src/ATen/native/Resize.cpp:19.)
  torch.add(a, b, out=d)
tensor([1, 2, 3, 4, 5])
torch.Size([5]) (1,)
```

**Comment by ngimel** [comment]
When user passes correctly sized discontiguous tensor, it should be directly used and its strides should be respected. This issue proposes disallowing incorrectly sized `out` tensors, so going forward there should be no question of passing incorrectly sized discontiguous tensor (which your example above is doing). For now, we reserve the right to do whatever we want with incorrectly sized `out` tensor.

**Comment by ngimel** [comment]
For future reference, from the offline conversation:
@ngimel:
Desired behavior is as follows (some functions, in particular TH/THC don’t respect it, for new functions we try to enforce it)
1. correctly sized output is respected. If function cannot produce the output with requested strides, intermediate copy is used
2. 0-sized output is resized and restrided to whatever striding is natural for the function to produce (that would be column-major for linalg functions)
3. incorrectly sized output warns currently, then is resized/restrided to the right size /natural striding for the op. Ideally we would like to deprecate it and error out in this case.

@mruberry
The UX contract with out= is that the operation performs its computation independent of the out= tensor and then (safe) copies its output to the out= tensor.
There are a couple caveats to this: we do resize/restride zero element tensors, and the copy is subject to device and dtype limitations.
This UX contract, however, should inform what operations can do.
Behind the scenes, however, functions can implement this behavior however they'd like. In particular, they can reuse the space provided by out= or fuse the copy operation

**Comment by mruberry** [comment]
I added a section to the Developer FAQ [here](https://github.com/pytorch/pytorch/wiki/Developer-FAQ#how-does-out-work-in-pytorch) describing this behavior.
